File: 2716-prime-subtraction-operation/prime-subtraction-operation.cpp
class Solution {
public:
    bool primeSubOperation(vector<int>& nums) {
        int maxElement = *max_element(nums.begin(), nums.end());

        // Store the sieve array.
        vector<int> sieve(maxElement + 1, 1);
        sieve[1] = 0;
        for (int i = 2; i <= sqrt(maxElement + 1); i++) {
            if (sieve[i] == 1) {
                for (int j = i * i; j <= maxElement; j += i) {
                    sieve[j] = 0;
                }
            }
        }

        // Start by storing the currValue as 1, and the initial index as 0.
        int currValue = 1;
        int i = 0;
        while (i < nums.size()) {
            // Store the difference needed to make nums[i] equal to currValue.
            int difference = nums[i] - currValue;

            // If difference is less than 0, then nums[i] is already less than
            // currValue. Return false in this case.
            if (difference < 0) {
                return 0;
            }

            // If the difference is prime or zero, then nums[i] can be made
            // equal to currValue.
            if (sieve[difference] == 1 or difference == 0) {
                i++;
                currValue++;
            } else {
                // Otherwise, try for the next currValue.
                currValue++;
            }
        }
        return 1;
    }
};

File: 2716-prime-subtraction-operation/prime-subtraction-operation.py
from math import isqrt

class Solution:
    def primeSubOperation(self, nums):
        max_element = max(nums)

        # Create sieve array for prime checking
        sieve = [1] * (max_element + 1)
        sieve[1] = 0
        for i in range(2, isqrt(max_element) + 1):
            if sieve[i] == 1:
                for j in range(i * i, max_element + 1, i):
                    sieve[j] = 0

        # Initialize curr_value to 1 and index to 0
        curr_value = 1
        i = 0

        while i < len(nums):
            # Calculate the difference needed to make nums[i] equal to curr_value
            difference = nums[i] - curr_value

            # If difference is less than 0, nums[i] is already less than curr_value
            if difference < 0:
                return False

            # If the difference is prime or zero, we can make nums[i] equal to curr_value
            if sieve[difference] == 1 or difference == 0:
                i += 1
                curr_value += 1
            else:
                # Otherwise, try for the next curr_value
                curr_value += 1

        return True
